wa/workloads/recentfling/__init__.py

```diff
@@ -0,0 +1,133 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+#pylint: disable=E1101,W0201
+
+import os
+import re
+
+from wa import Workload, Parameter, File
+from wa.framework.exception import WorkloadError, TargetError
+from wa.utils.exec_control import once
+
+
+class Recentfling(Workload):
+
+    name = 'recentfling'
+    description = """
+    Tests UI jank on android devices.
+
+    For this workload to work, ``recentfling.sh`` and ``defs.sh`` must be placed
+    in ``~/.workload_automation/dependencies/recentfling/``. These can be found
+    in the `AOSP Git repository <https://android.googlesource.com/platform/system/extras/+/master/tests/workloads>`_.
+
+    To change the apps that are opened at the start of the workload you will need
+    to modify the ``defs.sh`` file. You will need to add your app to ``dfltAppList``
+    and then add a variable called ``{app_name}Activity`` with the name of the
+    activity to launch (where ``{add_name}`` is the name you put into ``dfltAppList``).
+
+    You can get a list of activities available on your device by running
+    ``adb shell pm list packages -f``
+    """
+    supported_platforms = ['android']
+
+    parameters = [
+        Parameter('loops', kind=int, default=3,
+                  description="The number of test iterations."),
+        Parameter('start_apps', kind=bool, default=True,
+                  description="""
+                  If set to ``False``,no apps will be started before flinging
+                  through the recent apps list (in which the assumption is
+                  there are already recently started apps in the list.
+                  """),
+        Parameter('device_name', kind=str, default=None,
+                  description="""
+                  If set, recentfling will use the fling parameters for this
+                  device instead of automatically guessing the device.  This can
+                  also be used if the device is not supported by recentfling,
+                  but its screensize is similar to that of one that is supported.
+
+                  For possible values, check your recentfling.sh.  At the time
+                  of writing, valid values are: 'shamu', 'hammerhead', 'angler',
+                  'ariel', 'mtp8996', 'bullhead' or 'volantis'.
+                  """),
+    ]
+
+    @once
+    def initialize(self, context):  # pylint: disable=no-self-use
+        if self.target.get_sdk_version() < 23:
+            raise WorkloadError("This workload relies on ``dumpsys gfxinfo`` \
+                                 only present in Android M and onwards")
+
+        defs_host = context.resolver.get(File(self, "defs.sh"))
+        self.defs_target = self.target.install(defs_host)
+        recentfling_host = context.resolver.get(File(self, "recentfling.sh"))
+        self.recentfling_target = self.target.install(recentfling_host)
```

Comment:
As this is now only executed once for all instances of the workload, ``self.recentfling_target`` and ``self.defs_target`` should be set as class attributes i.e ``Recentfling.recentfling_target = ...``